### PR TITLE
fix: syslog plugin doesn't work

### PR DESCRIPTION
**Pull request by jiangfucheng**
### Description

Fixes https://github.com/apache/apisix/issues/9419

1.Convert the log message to RFC5424 format before sending it to the syslog server.
2.RFC5424 is “The Syslog Protocol", not only used for sls-logger. So, I relocated the file rfc5424.lua from apisix/plugins/slslog to apisix/syslog

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
@Sn0rt please help to review

**Comment by Sn0rt** [comment]
@jiangfucheng thx your PR.

can you add a test set to verify the syslog plugin is really working?

I think https://github.com/apache/apisix/blob/master/t/plugin/tcp-logger.t#L466 is helpful for you.

File: apisix/plugins/syslog/init.lua [hunk 1]
```diff
@@ -75,28 +79,32 @@ end
...
+    local rfc5424_data = rfc5424.encode("SYSLOG", "INFO", ctx.var.host,
+                                "apisix", ctx.var.pid, json_str)
+
+    if batch_processor_manager:add_entry(conf, rfc5424_data) then
         return
     end
 
     -- Generate a function to be executed by the batch processor
     local cp_ctx = core.table.clone(ctx)
     local func = function(entries, batch_max_size)
```
Comment (Sn0rt) [inline apisix/plugins/syslog/init.lua:97]: the `batch_max_size` unused

File: apisix/plugins/syslog/init.lua [hunk 2]
```diff
@@ -75,28 +79,32 @@ end
...
-        else
-            data, err = core.json.encode(entries) -- encode as array [{}]
-        end
-
-        if not data then
-            return false, 'error occurred while encoding the data: ' .. err
+        local items = {}
+        for _, e in ipairs(entries) do
+            table.insert(items, e)
+            core.log.info("buffered logs:", e)
```
Comment (Sn0rt) [inline apisix/plugins/syslog/init.lua:101]: maybe the log set with debug level is more suitable

**Comment by jiangfucheng** [comment]
@Sn0rt Updated, please review again, thanks.

**Comment by Sn0rt** [comment]
LGTM @monkeyDluffy6017 pls approve CI.

**Comment by monkeyDluffy6017** [comment]
@jiangfucheng Could you merge the master branch, some ci error are fixed

File: apisix/plugins/sls-logger.lua [hunk 3]
```diff
@@ -21,7 +21,7 @@ local bp_manager_mod = require("apisix.utils.batch-processor-manager")
 
 local plugin_name = "sls-logger"
 local ngx = ngx
-local rf5424 = require("apisix.plugins.slslog.rfc5424")
+local rf5424 = require("apisix.syslog.rfc5424")
```
Comment (leslie-tsang) [inline apisix/plugins/sls-logger.lua:24]: Please put the `syslog` library in the `plugins` folder. :)

File: apisix/plugins/sls-logger.lua [hunk 4]
```diff
@@ -138,9 +138,14 @@ function _M.log(conf, ctx)
...
     end
 
-    local rf5424_data = rf5424.encode("SYSLOG", "INFO", ctx.var.host,"apisix",
-                                      ctx.var.pid, conf.project, conf.logstore,
-                                      conf.access_key_id, conf.access_key_secret, json_str)
+    local structured_data = {
+        {name = "project", value = conf.project},
+        {name = "logstore", value = conf.logstore},
+        {name = "access-key-id", value = conf.access_key_id},
+        {name = "access-key-secret", value = conf.access_key_secret}
```
Comment (leslie-tsang) [inline apisix/plugins/sls-logger.lua:145]:
```suggestion
        {name = "access-key-secret", value = conf.access_key_secret},
```

File: apisix/plugins/syslog/init.lua [hunk 5]
```diff
@@ -18,6 +18,9 @@
 local core = require("apisix.core")
 local bp_manager_mod = require("apisix.utils.batch-processor-manager")
 local logger_socket = require("resty.logger.socket")
+local rfc5424 = require("apisix.syslog.rfc5424")
+local ipairs = ipairs
+local table = table
```
Comment (leslie-tsang) [inline apisix/plugins/syslog/init.lua:23]: please use the `core.table.*` instead. :)

File: apisix/plugins/syslog/init.lua [hunk 6]
```diff
@@ -75,28 +79,32 @@ end
...
-        else
-            data, err = core.json.encode(entries) -- encode as array [{}]
-        end
-
-        if not data then
-            return false, 'error occurred while encoding the data: ' .. err
+    local func = function(entries)
+        local items = {}
+        for _, e in ipairs(entries) do
+            table.insert(items, e)
```
Comment (leslie-tsang) [inline apisix/plugins/syslog/init.lua:100]:
```suggestion
            core.table.insert(items, e)
```
Would be better to `local` the `core.table.insert`.

File: apisix/syslog/rfc5424.lua [hunk 7]
```diff
@@ -95,10 +95,20 @@ function _M.encode(facility, severity, hostname, appname, pid, project,
...
+        for _, sd_param in ipairs(structured_data) do
+            structured_data_str = structured_data_str .. " " .. sd_param.name
+                                  .. "=\"" .. sd_param.value .. "\""
+        end
+        structured_data_str = structured_data_str .. "]"
+    end
+
+    return "<" .. pri .. ">1 " .. t .. " " .. hostname .. " "
+           .. appname .. " " .. pid .. " - " .. structured_data_str .. " "
+           .. msg .. "\n"
```
Comment (leslie-tsang) [inline apisix/syslog/rfc5424.lua:111]: please use `core.string.format` to format the output.

File: apisix/plugins/sls-logger.lua (hunk 3 above)
Comment (jiangfucheng) [reply apisix/plugins/sls-logger.lua:24]: The sls-logger also use `rfc5424.lua`, Is there any confusion in put `rfc5424.lua` to `plugins/syslog` folder?

Comment (jiangfucheng) [reply apisix/plugins/sls-logger.lua:24]: @leslie-tsang @Sn0rt @monkeyDluffy6017 Hi, could give me some suggestion?

Comment (monkeyDluffy6017) [reply apisix/plugins/sls-logger.lua:24]: The `apisix.syslog.rfc5424` is only used by plugin `sys-logger`, right?
So why not move the `apisix.syslog.rfc5424` to `apisix.plugins.slslog.rfc5424`?

Comment (jiangfucheng) [reply apisix/plugins/sls-logger.lua:24]: The `rfc5424.lua` both used by `syslog` and `slslogger`, so I moved the file to single other folder.

Comment (leslie-tsang) [reply apisix/plugins/sls-logger.lua:24]: > @leslie-tsang How about moving the file rfc5424 to `apisix/utils/` ?

LGTM

**Review by Sn0rt — DISMISSED** [review]

File: apisix/plugins/sls-logger.lua (hunk 3 above)
Comment (jiangfucheng) [reply apisix/plugins/sls-logger.lua]: Updated.

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by Sn0rt — APPROVED** [review]

**Review by soulbird — APPROVED** [review]

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**